### ISSUE TITLE
DyLoRA

### ISSUE THREAD

**Issue by space-nuko** [kohya-ss/sd-scripts#394]
https://arxiv.org/abs/2210.07558

Seems like there's code here already

https://github.com/huawei-noah/KD-NLP/tree/30eaa94c9dadea3d9af58de98f90a6ceb74bc234/DyLoRA

**Comment by mojivalipour** [comment]
@kohya-ss Hi, I just noticed that you tried to integrate our proposed method into your platform. First of all I would like to let you know that I appreciate your efforts. I skimmed over your code quickly and I just noticed that you overlooked into some part of the algorithm. In our experience we found these can be quite essential for stable convergence and better performance. 

As far as I understand, the scalar in your code is not same as ours: [Line 110](https://github.com/kohya-ss/sd-scripts/blob/ad5f318d066c52e5b27306b399bc87e41f2eef2b/networks/dylora.py#L110). We suggest to use $h= (W_0+\alpha/b \Delta W)X$ as we proposed in equation 7 in the [DyLoRA](https://arxiv.org/pdf/2210.07558.pdf) paper. It seems that you just fixed the scalar as $\alpha/r$ and multiplied that to the $\sqrt{r/b}$ which means as the result you will have $\alpha/r*\sqrt{r/b}$ which is different than what we proposed which it is $\alpha/b$. We needed $sqrt$ in our code as we used the same DyLoRA module as a wrapper over both $W_{dw}$ and $W_{up}$. 

Another observation that we had and I think it's going to help you is to use gradient accumulation. If you increase the gradient accumulation the performance might increase. You just need to be careful as increasing the gradient accumulation (depending to your implementation) might interfere with rank sampling procedure and bias the training toward some ranks. 

I also noticed you integrate the random rank selection during your forward pass [Line 77](https://github.com/kohya-ss/sd-scripts/blob/ad5f318d066c52e5b27306b399bc87e41f2eef2b/networks/dylora.py#L77) which means you will always change the rank randomly even during inference. Although it is possible but we suggest to give this possibility to the user to choose their desired rank at inference time. You only need the uniform or even geometric distribution during training.

**Comment by mojivalipour** [comment]
And for the frozen flag, only use it if you are looking for a sort of guarantee to avoid the already learned information in the lower ranks. This is not essential in practice so I don't suggest that. This is important to know as frozen performance is underperforming the non-frozen solution in an equal budget.  Finally, I am interested to see your results and the other users experience so please feel free to share your results with me so I might be able to better guide you.

**Comment by sdbds** [comment]
> @kohya-ss Hi, I just noticed that you tried to integrate our proposed method into your platform. First of all I would like to let you know that I appreciate your efforts. I skimmed over your code quickly and I just noticed that you overlooked into some part of the algorithm. In our experience we found these can be quite essential for stable convergence and better performance.
> 
> As far as I understand, the scalar in your code is not same as ours: [Line 110](https://github.com/kohya-ss/sd-scripts/blob/ad5f318d066c52e5b27306b399bc87e41f2eef2b/networks/dylora.py#L110). We suggest to use h=(W0+α/bΔW)X as we proposed in equation 7 in the [DyLoRA](https://arxiv.org/pdf/2210.07558.pdf) paper. It seems that you just fixed the scalar as α/r and multiplied that to the r/b which means as the result you will have α/r∗r/b which is different than what we proposed which it is α/b. We needed sqrt in our code as we used the same DyLoRA module as a wrapper over both Wdw and Wup.
> 
> Another observation that we had and I think it's going to help you is to use gradient accumulation. If you increase the gradient accumulation the performance might increase. You just need to be careful as increasing the gradient accumulation (depending to your implementation) might interfere with rank sampling procedure and bias the training toward some ranks.
> 
> I also noticed you integrate the random rank selection during your forward pass [Line 77](https://github.com/kohya-ss/sd-scripts/blob/ad5f318d066c52e5b27306b399bc87e41f2eef2b/networks/dylora.py#L77) which means you will always change the rank randomly even during inference. Although it is possible but we suggest to give this possibility to the user to choose their desired rank at inference time. You only need the uniform or even geometric distribution during training.

Hello, thank you for your great work. I have some questions about using Dylora with Optimizer Dadaptation

this is paper 
https://arxiv.org/abs/2301.07733
and code implement
https://github.com/facebookresearch/dadaptation

when i use dylora with it,it didn't work, so i make a [issue ](https://github.com/facebookresearch/dadaptation/issues/24) for this.
